Message List from Same User is now Distinguishable via Surface Tint on Long Press

Gaurav-Kushwaha-1225 opened the pull request:
## Changes Made

This issue resolved by:
- Wrapping the `MessageWithPossibleSender` with a `Container` 
- Then, giving it a color property on a condition that when `isMessageActionSheetOpen` is `true` show the surface tint, else remain `Colors.transparent`.

For this,
- I shifted the `MessageWithPossibleSender` from `StatelessWidget` to `StatefulWidget`.
- Added a nullable `VoidCallback? onDismiss` method in parameters of `showMessageActionSheet` or BottomSheet for updating the Surface Tint of message.

## Fixes #1142 

## Screenshots

| Light Theme | Dark Theme |
| -----------------| -------------------|
|             <video src="https://github.com/user-attachments/assets/717d5f1d-0368-4647-b570-dfa3359bfd1e" width=60%>                 |           <video src="https://github.com/user-attachments/assets/c2c6fb8a-5bbe-4280-a88e-c0de52e1fe62" width=60%> |

alya commented:
Please post screenshots of your changes in the PR description, as videos are hard to review.

Gaurav-Kushwaha-1225 commented:
> Please post screenshots of your changes in the PR description, as videos are hard to review.

## Changes

### 1st
- shifting the `MessageWithPossibleSender` widget from `StatelessWidget` to `StatefulWidget`.
- adding `isMessageActionSheetOpen` and `pressedTint` as parameters of `MessageWithPossibleSender` for making surface value tint to transparent after bottom sheet is closed.
- Screenshot: <img src="https://github.com/user-attachments/assets/395dc52b-972d-46cb-8c8d-85e215d8215a" width=80%>

### 2nd
- Wrapped with Container with color property on `isMessageActionSheetOpen` condition.
-  Updating the function of `onLongPress` of `MessageWithPossibleSender` widget. That, when `isMessageActionSheetOpen` is true show the surface tint. 
- Screenshot: <img src="https://github.com/user-attachments/assets/c12f5367-5c62-493f-aeea-5aa723fc411c" width=80%>

### 3rd
- Just adding a nullable `VoidCallback? onDismiss` method in parameters of `showMessageActionSheet` and `_showActionSheet` for getting response of closing of bottom sheet.
- Screenshots: 
- <img src="https://github.com/user-attachments/assets/68c73a14-5ff6-4904-898a-365fc35f2ad8" width=70%>
- <img src="https://github.com/user-attachments/assets/29d54238-f32a-4749-8119-08c8973301b0" width=70%>

chrisbobbe commented:
Alya means before/after screenshots of the app's UI. We can read your code easily without screenshots :)

chrisbobbe commented:
Also I see there's a failing check; please run tests locally (`tools/check`) and fix any failures.

Gaurav-Kushwaha-1225 commented:
> Alya means before/after screenshots of the app's UI. We can read your code easily without screenshots :)

I am sorry for the confusion.
Here, are the screenshots:

| Before | After |
| --------- | -------- |
| <img  src="https://github.com/user-attachments/assets/49ce3f5f-14d8-422f-95b4-178de071bfa2" > | <img src="https://github.com/user-attachments/assets/2a9df32e-bc5e-406a-b936-fbb9e390f629"> |
| <img  src="https://github.com/user-attachments/assets/ed0033c8-ac33-4dc3-a493-f0b5e22848fd" > | <img src="https://github.com/user-attachments/assets/7690186f-4e81-46c3-9a76-e1a0f1efe062"> |

> Also I see there's a failing check; please run tests locally (`tools/check`) and fix any failures.

I have updated the changes and there are no failing checks and conflicts now.
Thank you.

alya commented:
Are those before/after scheenshots the same? The point is to show the differences from your PR's changes.

Gaurav-Kushwaha-1225 commented:
@alya, 
In the screenshots, the message on which the user has long pressed, that message has appeared a surface tint color separating from rest other messages, in both light and dark theme.
This was the issue to resolve i.e. to give a surface tint to those messages where long press event has occurred.

alya commented:
Oh, I see it now.

Is it happening in the dark theme screenshot? I don't see a difference, so if the effect is there, it's too subtle.

alya commented:
It's also best to take screenshots that are identical other than the change your PR makes, to make them easy to compare.

Gaurav-Kushwaha-1225 commented:
> Oh, I see it now.
> 
> Is it happening in the dark theme screenshot? I don't see a difference, so if the effect is there, it's too subtle.

@alya 
Yeah, actually I took the same tint color for both themes. I saw in the figma file both are different for different themes. 
I will update the PR in few minutes.
Sorry, for my inconvenience.

alya commented:
Sure, please be mindful to review your own work carefully next time. Posting screenshots is one of the things that should help you identify any problems yourself, before asking someone else to take a look.

Gaurav-Kushwaha-1225 commented:
@alya 

Here, is the updated PR.

## Requirements according to Figma File
- <img  src="https://github.com/user-attachments/assets/72900580-413c-438a-b924-360442df64c7" width=70%> 
- <img src="https://github.com/user-attachments/assets/a5ecca4d-06cf-4657-8869-24674de3c982" width=70%> 

## Before and After Screenshots

> ### **_NOTE:_** In after screenshots, you may see the colors appearing to be a bit different (or less light/dark) from actual figma. This is because the action sheet is opened in my after screenshots which are not in figma file.

| Before | After |
| --------- | -------- |
| <img  src="https://github.com/user-attachments/assets/b698c4a2-43f8-472b-b5c7-83d533a1788e" > | <img src="https://github.com/user-attachments/assets/1f99492f-ca5f-481a-9009-6f6c7d29fe5f"> |
| <img  src="https://github.com/user-attachments/assets/a04369e2-7739-429e-9a49-eede25884cbe" > | <img src="https://github.com/user-attachments/assets/7d7f50eb-8d92-44ed-91a6-86d8d75634f6"> |

Both `pressedTint` colors in light and dark theme are according to the Figma File i.e. `RGBA(0, 0, 0, 0.04)` for light theme and `RGBA(1, 1, 1, 0.04)` for dark theme.

PIG208 commented:
Hi @Gaurav-Kushwaha-1225! Thanks for the updates. For the PR to be reviewable, we need tests for this: https://github.com/zulip/zulip-flutter?tab=readme-ov-file#submitting-a-pull-request

Please also pay attention to the point on commit style guide from that link, and organize the commit history coherently.

chrisbobbe commented:
I see you're adding merge commits; Zulip doesn't use those.

Gaurav-Kushwaha-1225 commented:
> Hi @Gaurav-Kushwaha-1225! Thanks for the updates. For the PR to be reviewable, we need tests for this: https://github.com/zulip/zulip-flutter?tab=readme-ov-file#submitting-a-pull-request
> 
> Please also pay attention to the point on commit style guide from that link, and organize the commit history coherently.

Hi @PIG208, I have added the necessary tests for these changes in `test/widgets/message_list_test.dart` file.
Please review the changes and feature added. 
Thank you.

PIG208 commented:
Looks like this update still hasn't addressed my previous comment on the commit style:

> Please also pay attention to the point on commit style guide from that link, and organize the commit history coherently.

As Chris said, we don't use merge commits. Could you please also review the commit style guide linked in our [README](https://github.com/zulip/zulip-flutter?tab=readme-ov-file#submitting-a-pull-request)?

Gaurav-Kushwaha-1225 commented:
> Looks like this update still hasn't addressed my previous comment on the commit style:
> 
> > Please also pay attention to the point on commit style guide from that link, and organize the commit history coherently.
> 
> As Chris said, we don't use merge commits. Could you please also review the commit style guide linked in our [README](https://github.com/zulip/zulip-flutter?tab=readme-ov-file#submitting-a-pull-request)?

I’ve updated the commits, improving the commit style and making other enhancements. 
I believe it’s ready for review now. 

Please let me know if any further changes are required.

chrisbobbe commented on lib/widgets/message_list.dart:
```diff
@@ -1347,40 +1354,63 @@ class MessageWithPossibleSender extends StatelessWidget {
-      onLongPress: () => showMessageActionSheet(context: context, message: message),
+      onLongPress: () async {
+        setState(() {
+          isMessageActionSheetOpen = true;
```
If `isMessageActionSheetOpen` is supposed to mean what its name says, it isn't accurate here. The action sheet won't be open until 100ms have passed.

I'd like to try a different approach, using an existing Flutter API that seems suitable.

- `_MessageWithPossibleSenderState` gets a new final field: 
  ```dart
    final WidgetStatesController statesController = WidgetStatesController();
  ```

- Override `_MessageWithPossibleSenderState.initState` like so:

  ```dart
    @override
    void initState() {
      super.initState();
      statesController.addListener(() {
        setState(() {
          // Force a rebuild to resolve background color
        });
      });
    }
  ```

- Override `_MessageWithPossibleSenderState.dispose`, calling `statesController.dispose()`
- In addition to the existing `onLongPress`, pass `onLongPressDown`, `onLongPressCancel`, and `onLongPressUp` to the `GestureDetector`. Each passed callback should call `statesController.update(WidgetState.pressed, …` with true or false as appropriate.
- At the top of the existing `onLongPress` calback, call `statesController.update(WidgetState.selected, true)`. When the action sheet is dismissed, unset that with `false`.
- For the background color, use:

  ```dart
              color: const WidgetStateColor.fromMap({
                WidgetState.pressed: designVariables.pressedTint,
                WidgetState.selected: designVariables.pressedTint,
                WidgetState.any: Colors.transparent,
              }).resolve(statesController.value)
  ```

chrisbobbe commented on lib/widgets/message_list.dart:
```diff
@@ -1347,40 +1354,63 @@ class MessageWithPossibleSender extends StatelessWidget {
+          isMessageActionSheetOpen = true;
+        });
+
+        await Future.delayed(const Duration(milliseconds: 100), () {
```
I don't think we want an additional 100ms delay here; let's remove it.

chrisbobbe reviewed (changes requested):
Thanks! Comments below. Please also squash these two commits into one, so the tests land in the same commit as the code they're testing.

gnprice commented on lib/widgets/action_sheet.dart:
```diff
@@ -61,7 +62,7 @@ void _showActionSheet(
               const ActionSheetCancelButton(),
             ])));
-    });
+    }).whenComplete(onDismiss ?? () {});
```
This type of API based on callbacks can make the control flow pretty confusing to follow. The behavior here can be more cleanly modeled using Future.

Let's also avoid using explicit Future methods like `whenComplete`, in favor of `async` and `await`.

Concretely, let's have this function `_showActionSheet`, and its caller `showMessageActionSheet`, mimic the API we developed for `showErrorDialog`:
```dart
/// Displays an [AlertDialog] with a dismiss button.
///
/// The [DialogStatus.closed] field of the return value can be used
/// for waiting for the dialog to be closed.
// This API is inspired by [ScaffoldManager.showSnackBar].  We wrap
// [showDialog]'s return value, a [Future], inside [DialogStatus]
// whose documentation can be accessed.  This helps avoid confusion when
// intepreting the meaning of the [Future].
DialogStatus showErrorDialog({
```

We can use the same `DialogStatus` type for the return value. I'll update your PR branch with a quick prep commit that gives it a slightly more general name `ModalStatus`, and slightly more general docs, so that it's ready to be used that way.

gnprice reviewed (commented):
In addition to Chris's comments above, one other high-level comment. (I've only done an initial skim of this PR, since it's still in maintainer review.)

gnprice commented:
> I'll update your PR branch with a quick prep commit

Done. @Gaurav-Kushwaha-1225 As the first step when you start work on your next revision, be sure to pull down the updated version of this branch — use `git fetch me` or `git fetch origin`, whatever name you use for the remote pointing at your GitHub fork of this repo. (If you're unsure how to do this, the `#git help` channel is a good place for questions.)

Gaurav-Kushwaha-1225 commented:
Hi @chrisbobbe, I have updated the changes as per your request.

- Changed return type of `showMessageActionSheet` and `_showActionSheet` to `ModalStatus`
- Used the logic as you said in above requests for `pressedTint`

https://github.com/user-attachments/assets/a9e36caa-5403-459d-b1fa-a01f4266bbc1

chrisbobbe commented on lib/widgets/theme.dart:
```diff
@@ -327,6 +330,7 @@ class DesignVariables extends ThemeExtension<DesignVariables> {
   final Color modalBarrierColor;
   final Color mutedUnreadBadge;
   final Color navigationButtonBg;
+  final Color pressedTint;
```
Bump on https://github.com/zulip/zulip-flutter/pull/1152#discussion_r1919332516 : this doesn't belong in this section, which has this description:

```dart
  // Not named variables in Figma; taken from older Figma drafts, or elsewhere.
```

This _is_ a named variable in Figma, as it seems you've noticed 🙂

https://github.com/zulip/zulip-flutter/pull/1152#discussion_r1920456371
> It is named as `pressedTint` itself in the design file.

so it goes in the same section as other named Figma variables, so (in current `main`) between `mainBackground` and `textInput`.

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1096,6 +1097,79 @@ void main() {
+      late Message message;
+
+      setUp(() {
+        // Create a basic message for testing
```
This comment is just restating what the code does; let's remove the comment.

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1096,6 +1097,79 @@ void main() {
+      testWidgets('starts with transparent background', (tester) async {
+        await buildTestWidget(tester);
+
+        expect(
```
We use `check`, not `expect`; see existing tests

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1096,6 +1097,79 @@ void main() {
+        return (decoratedBox.decoration as BoxDecoration).color;
+      }
+
+      Future<void> buildTestWidget(WidgetTester tester) async {
+        await setupMessageListPage(
+          tester,
+          messages: [message],
+        );
+      }
```
How about just calling `setupMessageListPage` in each tests; I don't think we need this helper function.

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1096,6 +1097,79 @@ void main() {
+        await tester.longPress(find.byType(MessageWithPossibleSender));
+        await tester.pump();
+
+        final expectedTint = DesignVariables.of(tester.element(find.byType(MessageWithPossibleSender)))
+            .pressedTint;
```
indentation nit: two-space indent, like elsewhere

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1096,6 +1097,79 @@ void main() {
+      testWidgets('shows tint color when long pressed', (tester) async {
+        await buildTestWidget(tester);
+
+        // Trigger long press
+        await tester.longPress(find.byType(MessageWithPossibleSender));
```
Another comment that just restates what's clearly expressed in the code 🙂; let's remove it

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1096,6 +1097,79 @@ void main() {
+        );
+      });
+
+      testWidgets('returns to transparent after action sheet dismissal', (tester) async {
```
We should also check that the message is tinted when it's not being long-pressed but the action sheet hasn't been dismissed yet.

Gaurav-Kushwaha-1225 replied:
```dart
testWidgets('maintains tint color while action sheet is open', (tester) async {
  await setupMessageListPage(tester, messages: [message]);

  // Perform long press to open the action sheet
  await tester.longPress(find.byType(MessageWithPossibleSender));
  await tester.pump();

  final expectedTint = DesignVariables.of(tester.element(find.byType(
    MessageWithPossibleSender))).pressedTint;

  // Verification to ensure the tint persists
  await tester.pump(const Duration(milliseconds: 500));

  check(getBackgroundColor(tester),
    because: 'Message should continue to show tint color while action sheet is visible',
  ).equals(expectedTint);
});
```

Is this piece of code fine for this testing what you have just said above?

Since, it is very similar to one of the previous tests `shows tint color when long pressed`, except here that we are waiting for 500 milliseconds before checking its color.

chrisbobbe replied:
That looks reasonable. One small comment is that the code comments are unnecessary (the code is clear enough without them) so let's not add them.

chrisbobbe commented on lib/widgets/message_list.dart:
```diff
@@ -1347,40 +1370,61 @@ class MessageWithPossibleSender extends StatelessWidget {
-                  : null),
-            ]),
-        ])));
+        child: DecoratedBox(
```
Let's actually swap this `DecoratedBox` with its parent `Padding`. That way, the tint also applies on the padding at the top and bottom, like in the [Figma](https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=3988-34789&m=dev).

chrisbobbe reviewed (commented):
Thanks! Small comments below.

Gaurav-Kushwaha-1225 commented:
Hi @chrisbobbe, 
I have updated this PR according to your review.

Please have a look at it and let me know if anything else is required.
Thank you! 😊

Gaurav-Kushwaha-1225 commented:
Hi @chrisbobbe,
I have updated this feat. PR accordingly.

Please take a review and let me know if something else is required.

Gaurav-Kushwaha-1225 commented:
Hi @chrisbobbe,
I have updated this feat. PR accordingly.

Please take a review and let me know if something else is required.

Gaurav-Kushwaha-1225 commented:
## Screenshots:

Below are the screenshots of added `pressedTint` feature. Unlike in previous screenshots of this PR, the outer padding is also included now as asked in above review comments.
| Before | After |
| --------- | -------- |
| <img  src="https://github.com/user-attachments/assets/2ca9b7a6-f3b7-4ce2-ad49-09b2b0fe51b2" > | <img src="https://github.com/user-attachments/assets/bfa1e3ce-7a9c-4a87-ae58-585af23a473f"> |
| <img  src="https://github.com/user-attachments/assets/b6e75979-0205-47fa-8ffb-f82c879eda2c" > | <img src="https://github.com/user-attachments/assets/5f2a4c51-3504-4db8-bf5e-c13d26a0c578"> |

gnprice commented:
Thanks @Gaurav-Kushwaha-1225 and @chrisbobbe for your work on this so far!

The issue this addresses is an M6 Post-launch issue, and we're now [focusing more strictly on launch issues](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/Flutter.20beta.20timeline/near/2078556) until the launch is ready. But given that a substantial amount of review work has gone into this and it seems like it's relatively close to reaching a state we're able to merge, this can be one of the exceptions that we continue work on nonetheless.

chrisbobbe commented:
Hi @Gaurav-Kushwaha-1225, GitHub is saying there are some merge conflicts; would you resolve those please?

Gaurav-Kushwaha-1225 commented:
> Hi @Gaurav-Kushwaha-1225, GitHub is saying there are some merge conflicts; would you resolve those please?

I've updated the commits, and there are no more merge conflicts. However, the current failing CI checks are due to a separate issue in `tools/check-android`, as discussed in [Zulip Chat](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/Gradle.20out.20of.20memory/with/2076887).

chrisbobbe commented:
Thanks! Ah, but I see now there are some new merge conflicts; would you resolve those please?

Gaurav-Kushwaha-1225 commented:
> I've updated the commits, and there are no more merge conflicts.

I've updated the commits, and there are no more merge conflicts. 
Hopefully, there won't be any conflicts when you review the PR again. 😓

chrisbobbe commented on lib/widgets/content.dart:
```diff
@@ -1318,12 +1318,12 @@ class MessageTableCell extends StatelessWidget {
 
 void _launchUrl(BuildContext context, String urlString) async {
-  DialogStatus showError(BuildContext context, String? message) {
+  ModalStatus showError(BuildContext context, String? message) {
```
```
dialog [nfc]: Generalize to ModalStatus, renaming DialogStatus
```

This looks like the only change in this file that's intended—the others look like regressions to an older `main` (like reverting some UI strings to not be translatable), probably accidents in resolving merge conflicts. Greg's thoughts here will be useful to you: https://github.com/zulip/zulip-flutter/pull/1017#discussion_r1955567044

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1188,6 +1189,77 @@ void main() {
     });
+
+
+    group('action sheet visual feedback', () {
```
The visual feedback is also given on long-press gestures that are canceled and don't cause the action sheet to open. How about just 'background-color tint'?

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1188,6 +1189,77 @@ void main() {
 
       debugNetworkImageHttpClientProvider = null;
     });
+
```
nit: only add one empty line, not two

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1188,6 +1189,77 @@ void main() {
+        await setupMessageListPage(tester, messages: [message]);
+
+        check(getBackgroundColor(tester),
+          because: 'Message should start with transparent background',
```
The `because` isn't needed; the test description is sufficient. Let's remove the `because`. (Here and in the other added tests.)

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1188,6 +1189,77 @@ void main() {
+        return (decoratedBox.decoration as BoxDecoration).color;
+      }
+
+      testWidgets('starts with transparent background', (tester) async {
```
Ah, there's another important case that we should test that isn't covered yet: the `onLongPressCancel` case. The user puts their finger down, then quickly drags because they want to scroll instead of opening the action sheet. The UI recognizes this as cancelling the long-press gesture, and we should test that the feedback disappears at that time.

Rather than just adding another `testWidgets`, I think a neat way to organize the test coverage is in just two tests, each with a clear storyline:

- 'long-press opens open action sheet':
  - Check transparent
  - Begin long-press
  - Check tinted
  - End long-press
  - Check tinted (it didn't flicker to transparent here)
  - wait through the bottom-sheet entrance animation (for examples, search for 'default duration of bottom-sheet enter animation')
  - Check tinted
  - Dismiss action sheet
  - Check transparent
- 'long-press canceled':
  - Check transparent
  - Begin long-press
  - Check tinted
  - Cancel long-press, for example by dragging
  - Check transparent

Gaurav-Kushwaha-1225 replied:
Hi @chrisbobbe,

The above comment https://github.com/zulip/zulip-flutter/pull/1152#discussion_r1956734311 and this comment are referring to the same issue?

If yes, then np. 
I have updated the PR with both the required tests and changes.

Thank You.

chrisbobbe reviewed (commented):
Thanks for the revision! Here's another review, this time with a new thought about a case that needs test coverage.

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -1226,6 +1227,67 @@ void main() {
+        check(getBackgroundColor(tester)).equals(expectedTint);
+
+        await gesture.moveBy(const Offset(0, 50));
+        await gesture.up();
+        await tester.pumpAndSettle();
```
Can we comment out the `gesture.up`? I believe the drag by itself (without lifting a finger) should suffice, and simplifies the storyline.

Also can the `pumpAndSettle` just be a `pump`?

chrisbobbe reviewed (commented):
Thanks! Comments below, including some substantive comments on the tests.

Gaurav-Kushwaha-1225 commented on test/widgets/message_list_test.dart:
```diff
@@ -1226,6 +1227,67 @@ void main() {
+        return (decoratedBox.decoration as BoxDecoration).color;
+      }
+
+      testWidgets('long-press opens action sheet', (tester) async {
+        await setupMessageListPage(tester, messages: [message]);
+
+        check(getBackgroundColor(tester)).equals(Colors.transparent);
+
+        await tester.longPress(find.byType(MessageWithPossibleSender));
+        await tester.pump();
+
+        final expectedTint = DesignVariables.of(tester.element(find.byType(MessageWithPossibleSender)))
+          .pressedTint;
+
+        check(getBackgroundColor(tester)).equals(expectedTint);
+
+        await tester.pumpAndSettle();
+
+        check(getBackgroundColor(tester)).equals(expectedTint);
+
+        await tester.tapAt(const Offset(0, 0));
+        await tester.pumpAndSettle();
+
+        check(getBackgroundColor(tester)).equals(Colors.transparent);
+      });
```
According to the above comments, I think this would be the updated version. I also added comments so that I can describe what I think is happening via those respective lines.

```dart
  testWidgets('long-press opens action sheet', (tester) async {
    await setupMessageListPage(tester, messages: [message]);

    // Check starts as transparent
    check(getBackgroundColor(tester)).equals(Colors.transparent);

    // Begin long-press - part (a)
    final gestureFuture = tester.longPress(find.byType(MessageWithPossibleSender));
    await tester.pump(const Duration(milliseconds: 300));

    final expectedTint = DesignVariables.of(tester.element(find.byType(MessageWithPossibleSender)))
      .pressedTint;

    // Check tinted after short pump
    check(getBackgroundColor(tester)).equals(expectedTint);

    // Complete long-press
    await tester.pump(const Duration(milliseconds: 300));
    await gestureFuture;

    // Check tinted after long-press 
    // (part b - Not yet completed, as it is not failing a particular case when implemented that it should fail.)
    check(getBackgroundColor(tester)).equals(expectedTint);

    // Wait explicitly for bottom-sheet entrance animation (250ms)
    await tester.pump(const Duration(milliseconds: 250));

    // Check still tinted
    check(getBackgroundColor(tester)).equals(expectedTint);

    // Dismiss action sheet
    await tester.tapAt(const Offset(0, 0));
    await tester.pumpAndSettle();

    // Check returns to transparent
    check(getBackgroundColor(tester)).equals(Colors.transparent);
  });
```

## Breakdown of Changes

### **1. For part (a)**

- It's now implemented as requested.
- Used **300 milliseconds** because `await pump(kLongPressTimeout + kPressTimeout);` summed up to **600ms**.
- However, an error occurs while running the test because `await` should be there after a future method call:

```bash
The following assertion was thrown running a test:
Guarded function conflict.
You must use "await" with all Future-returning test APIs.
The guarded method "longPressAt" from class WidgetController was called from
Then, the "pump" method from class WidgetTester was called from
The first method (WidgetController.longPressAt) had not yet finished executing at the time that the
second method (WidgetTester.pump) was called. Since both are guarded, and the second was not a
nested call inside the first, the first must complete its execution before the second can be called.
Typically, this is achieved by putting an "await" statement in front of the call to the first.
```

### **2. For part (b)**

- I completely understood the request, butt.
- This check isn't failing, but after making the suggested changes (in details part [discussion link](https://github.com/zulip/zulip-flutter/pull/1152#discussion_r1968481632)), as it **should** fail.
- I am trying to figure out where I am missing in this case of checking tint after long press.

### **3. For part (c)**

- Used **250ms** for waiting, as I couldn't find a default duration for bottom-sheet enter animation.
- The Flutter documentation states that the default value is **null**.
- References:
  - [Bottom Sheet Animation Controller](https://api.flutter.dev/flutter/material/BottomSheet/createAnimationController.html)
  - [Bottom Sheet Animation Controller Constructor](https://api.flutter.dev/flutter/animation/AnimationController/AnimationController.html)

I will also ask regarding part (b) in the `mobile-dev-help` group for help.
Thank You.

Gaurav-Kushwaha-1225 commented:
Today, I noticed an issue in this feature implementation, occurring in both versions of the last commit—whether I was using `GestureDetector` or the current implementation with `RawGestureDetector` in `lib/widgets/message_list.dart`.

Video Demonstration:
<details>

[Bug Video Demo](https://github.com/user-attachments/assets/47541901-b0fe-494e-9e76-cc4b5469c513)

</details>

This might be due to the parent widget, which is of type `ScrollController`. When scrolling up, it also detects a long press event, leading to this unintended behavior.

Solution:  
After investigating the issue, I found a solution: wrapping the child widget with a `Listener` widget.  

```dart
return RawGestureDetector(
  behavior: HitTestBehavior.translucent,
  ...
  ...
  ...
  child: Listener(
    onPointerDown: (_) {
      statesController.update(WidgetState.pressed, true);
    },
    onPointerUp: (_) {
      statesController.update(WidgetState.pressed, false);
    },
    onPointerCancel: (_) {
      statesController.update(WidgetState.pressed, false);
    },
    child: DecoratedBox(
      decoration: BoxDecoration(
        color: WidgetStateColor.fromMap({
          WidgetState.pressed: designVariables.pressedTint,
          WidgetState.selected: designVariables.pressedTint,
          WidgetState.any: Colors.transparent,
        }).resolve(statesController.value))
```

Video after Implementing above Solution:
<details>

[Video after implementing Solution](https://github.com/user-attachments/assets/64f1218b-2c32-42f2-9626-a7bd6a9fad86)

</details>

gnprice commented:
This is a post-launch issue, and at https://github.com/zulip/zulip-flutter/pull/1152#issuecomment-2638808997 last month I decided we'd make an exception to continue work on it before launch because it looked like it was close to merge. From the thread since then, though, it seems like it wasn't yet close, and it's not clear to me that it's close to merge now either.

So let's pause work on this PR like we have on other PRs for post-launch issues. We can return to it after the app is launched in the coming months. Thanks again @Gaurav-Kushwaha-1225 for all your work on it.
